### PR TITLE
Avoid listing all schemas for USE in Iceberg

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/hms/TestTrinoHiveCatalogWithHiveMetastore.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    // Use MinIO for storage, since HDFS is hard to get working in a unit test
+    private HiveMinioDataLake dataLake;
+
+    @BeforeClass
+    public void setUp()
+    {
+        dataLake = new HiveMinioDataLake(
+                bucketName,
+                HiveHadoop.HIVE3_IMAGE);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/hms/TestTrinoHiveCatalogWithHiveMetastore.java:67]: nit: Static import `HIVE3_IMAGE`. I would change to single line as it's very simple.

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestTrinoHiveCatalog.java [hunk 2]
```diff
@@ -38,15 +39,15 @@
...
 
-    public TestTrinoHiveCatalog()
-            throws IOException
+    @BeforeClass
+    public void setUp()
+            throws Exception
     {
         tempDir = Files.createTempDirectory("test_trino_hive_catalog");
-        metastoreDir = tempDir.resolve("iceberg_data").toFile();
+        File metastoreDir = tempDir.resolve("iceberg_data").toFile();
```
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestTrinoHiveCatalog.java]: yes

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 3]
```diff
@@ -206,6 +207,14 @@ public FileHiveMetastore(NodeVersion nodeVersion, HdfsEnvironment hdfsEnvironmen
     public synchronized void createDatabase(Database database)
     {
         requireNonNull(database, "database is null");
+        database = new Database(
```
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:210]: i want to make sure all fields are copied. using constructor ensures code is updated, using a builder does not.

**Comment by findepi** [comment]
AC

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java [hunk 4]
```diff
@@ -202,7 +202,7 @@ public void dropNamespace(ConnectorSession session, String namespace)
             try {
                 return !fileSystemFactory.create(session).listFiles(path).hasNext();
             }
-            catch (IOException e) {
+            catch (IOException | RuntimeException e) {
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java:219]: cc @jirassimok

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 5]
```diff
@@ -64,7 +65,14 @@
...
     {
-        return listSchemaNames(session).contains(schemaName);
+        if (!schemaName.equals(schemaName.toLowerCase(ENGLISH))) {
+            // Currently, Trino schemas are always lowercase, so this one cannot exist (https://github.com/trinodb/trino/issues/17)
+            return false;
+        }
+        return listSchemaNames(session).stream()
+                // Lower-casing is done by callers of listSchemaNames (see MetadataManager)
+                .map(schema -> schema.toLowerCase(ENGLISH))
+                .anyMatch(schemaName::equals);
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:75]: cc @martint

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java (hunk 4 above)
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java:219]: In the commit comment:

```
The code was handling `IOException`, but getting a filesystem mail also
fail ...
```

`mail` -> `may` typo?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 6]
```diff
@@ -290,6 +290,12 @@ public IcebergMetadata(
         this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
     }
 
+    @Override
+    public boolean schemaExists(ConnectorSession session, String schemaName)
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:294]: A follow-up for Delta & Hive would be welcomed:
`io.trino.plugin.deltalake.metastore.HiveMetastoreBackedDeltaLakeMetastore#getDatabase`
`io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore#getDatabase`

**Review by findinpath — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java [hunk 7]
```diff
@@ -48,40 +58,88 @@
...
     {
         TrinoCatalog catalog = createTrinoCatalog(false);
-        Path tmpDirectory = Files.createTempDirectory("iceberg_catalog_test_create_table_");
-        tmpDirectory.toFile().deleteOnExit();
 
+        String namespace = "testNonLowercaseNamespace" + randomTableSuffix();
+        // Trino schema names are always lowercase (until https://github.com/trinodb/trino/issues/17)
+        String schema = namespace.toLowerCase(ENGLISH);
+
+        // Currently this is actually stored in lowercase by all Catalogs
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java:91]: Is that true if you create the namespace via Spark?

Comment (findepi) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java:91]: that's a property of the Glue/HMS, not the engine.
Note that TrinoCatalog implementations don't lowercase themselves

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 6 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:294]: discussed offline; it's unclear whether worth it

**Pull request merged**